### ISSUE TITLE
Calculation switches

### ISSUE THREAD

**Issue by shaunhaskey** [D3DEnergetic/FIDASIM#72]
@lstagner @bgrierson21 From messing around, I think I found out that if I only want to calculate the thermal D-alpha spectrum due to dcx and halo (to maximise the speed of the calculation I am interested in) I need to set the following switches (if calc_bes isn't selected then the code gets angry): 

calc_npa= 0
calc_brems= 0
calc_bes= 1
calc_fida= 0
calc_birth= 0
calc_fida_wght= 0
calc_npa_wght= 0

Is the correct? Is there a dedicated switch for the thermal DCX and halo D-alpha emission or is that done by default if one of hte above switches (i.e calc_bes) is selected?

Should I continue to post questions on github instead of emailing them?

Thanks

**Comment by bgrierson21** [comment]
Luke will know the answer.  In the past it was all lumped into one spectrum, and if this is the case now, then we can still use it.  But eventually being able to record the direct and subsequent emission sources is highly desirable.

**Comment by lstagner** [comment]
Yes, that is the correct setup for what you want. 

`calc_bes` is the catch all switch for calculating the Full, Half, Third, and Halo spectra. I decided against splitting it up because most of the time users would want Beam and Halo spectra at the same time. Also It didn't make much sense to separate out the Halo switch because it would require the beam neutrals to be calculated anyway so might as well just do it all. 

@bgrierson21 There is a `dump_dcx` switch that will dump the dcx neutrals and spectra to file. If you want every single Halo generation then that could also be done with some work.

**Comment by shaunhaskey** [comment]
Thanks!

**Comment by bgrierson21** [comment]
Lets make this clear.  The "halo", full, half and third energy components are stored in different arrays, correct?  We should never force the user to receive the "complete" spectrum as a single array without separating the parts.

**Comment by lstagner** [comment]
Yes. Separate arrays. The file looks like this
```
IDL> a=read_hdf5("/p/fida/lstagner/TEST/test_1a_spectra.h5")
% Compiled module: READ_HDF5.
% Loaded DLM: HDF5.
IDL> help,a
** Structure <9951018>, 12 tags, length=304360, data length=304352, refs=1:
   BREMS           STRUCT    -> <Anonymous> Array[1]
   FIDA            STRUCT    -> <Anonymous> Array[1]
   FULL            STRUCT    -> <Anonymous> Array[1]
   HALF            STRUCT    -> <Anonymous> Array[1]
   HALO            STRUCT    -> <Anonymous> Array[1]
   LAMBDA          STRUCT    -> <Anonymous> Array[1]
   NCHAN           STRUCT    -> <Anonymous> Array[1]
   NLAMBDA         STRUCT    -> <Anonymous> Array[1]
   RADIUS          STRUCT    -> <Anonymous> Array[1]
   THIRD           STRUCT    -> <Anonymous> Array[1]
   VERSION         STRING    'v0.3.2-9-g9973b03'
   DESCRIPTION     STRING    'Spectra calculated by FIDASIM'

IDL> help,a.full
** Structure <9966cd8>, 3 tags, length=48032, data length=48032, refs=2:
   DESCRIPTION     STRING    'Full energy component of the beam emmision: full(lambda,chan)'
   UNITS           STRING    'Ph/(s*nm*sr*m^2)'
   DATA            DOUBLE    Array[2000, 3]

```

**Comment by bgrierson21** [comment]
Agreed.  That is good.

**Comment by lstagner** [comment]
> (if calc_bes isn't selected then the code gets angry): 
Angry in what way?

**Comment by shaunhaskey** [comment]
Crashes, but all of the calc_ switches were set to 0 so not very surprising. I was trying to see if I could do that and still get the spectrum from DCX and halo. Clearly I can't. Do you think it would be a good idea to include another entry in the input file that defines how many generations of halo's will be calculated? This could add a bit more of a speed saving for simulating the main ion CER spectrum.... although the iteratinos do get faster as the number of particles drops.

**Comment by lstagner** [comment]
It shouldn't have crashed. I fixed the problem in #77.

>Do you think it would be a good idea to include another entry in the input file that defines how many generations of halo's will be calculated?

Not really. If it where a user defined quantity I am afraid that a user will set it and forget it. Eventually there will be a run that actually requires more iterations and user wont realize. No. I think it is safer just to let the physics decide.

In regards to making the code faster (your getting a bit greedy) I haven't actually profiled the code yet. If we ever did that then I am sure there are places where we can trim the fat so to speak.
